Fix performance for My Courses shortcode.

aaronfc opened the pull request:
Fixes #5247

### Changes proposed in this Pull Request
* Tweak the query arguments for courses to avoid querying the full list of users associated to the course when not needed.
* Reimplemented the `has_stored_enrolment` method for the same reason.
* More details in the P2 post: p6rkRX-3Kn-p2

### Testing instructions
* Test that everything is working as expected after this change.
* Test that My Courses page loads faster.
  * In order to reproduce this you will need to have many users registered in your local site... I did it by using All In One Migration plugin from wpcourses to my local site, but I spent multiple hours with this).

<!--
Helpful tips for screenshots:
https://en.support.wordpress.com/make-a-screenshot/
-->
### Screenshot / Video
#### BEFORE
<img width="1727" alt="image" src="https://user-images.githubusercontent.com/799065/176167107-41c7f2a7-8ebd-4e96-9995-eb77f2173b67.png">

#### AFTER
<img width="1728" alt="image" src="https://user-images.githubusercontent.com/799065/176166897-b24ba01c-25d3-47ed-ad9a-3746d781100f.png">

aaronfc committed:
Disable caching for enroled courses query.

Caching also makes terms cache to be calculated and saved, which is a
heavy task when a high amount of users are enroled.

aaronfc committed:
Disable cache when checking if user is enroled.

Checking the cache also triggers the calculation for all the users which
is a heavy task for courses with high amount of users enroled.

aaronfc committed:
Remove save_enrolment optimisation.

In order to do this optimisation we had to call `has_term` which in the
end calculates all the terms for the course. This has a heavy cost on
courses with high amount of users enroled.

aaronfc committed:
Reimplement `has_stored_enrolment` to reverse logic.

For performance reasons it is better to query for the posts (courses)
associated to the term (specific user) and not the other way around.
Previous implementation queried for all the terms for a given course,
which is a heavy query.

aaronfc commented on includes/class-sensei-learner.php:
```diff
@@ -377,6 +377,8 @@ public function get_enrolled_courses_query_args( $user_id, $base_query_args = []
 			'order'       => $order,
 			'orderby'     => $orderby,
 			'tax_query'   => [], // phpcs:ignore WordPress.DB.SlowDBQuery.slow_db_query_tax_query -- Just empty to set array.
+			'lazy_load_term_meta' => false,
+			'cache_results' => false,
```
Without this the post is cached and information about the associated terms too; implying a query for all the terms associated to all the taxonomies in the post.

aaronfc committed:
Revert "Disable cache when checking if user is enroled."

This reverts commit 9dc7f444d35edf879d0cad416bcae1a07f851744.
Which is not needed anymore after reimplementing the
`has_stored_enrolment` method.

aaronfc committed:
Remove extra new line.

aaronfc committed:
Fix indentation.

aaronfc committed:
Add argument comment.

aaronfc committed:
Fix in_array comparison by casting to string.

renatho reviewed (commented):
I didn't migrate the whole wpcourses site to test, but it seems to be working well, and it seems safe.

I'm also adding @gkaragia as a reviewer since he worked a lot in the enrollment implementation, just to confirm if he didn't see anything that I could have missed.

gikaragia commented:
Although I couldn't test them, the changes look good and make sense to me! Just a couple of questions:
- Is there any improvement if instead of setting `lazy_load_term_meta` we set `$update_post_term_cache` to false?
- Does this mean that we don't have a big slow query slowing down the side, but we have thousands of small ones? Is it possible to share your profiler findings? I think that they might be useful for the future.

aaronfc commented:
Hi @gkaragia, thanks for the review!

>Is there any improvement if instead of setting lazy_load_term_meta we set $update_post_term_cache to false?

I tried and compared to settinng `lazy_load_term` it worsened a little (not too much). Response time increased slightly and the total amount of calls for the request increased from 1,102,980 to 1,120,241. For the moment I am letting `lazy_load_term` only. Let me know if you want me to try anything else.

> Does this mean that we don't have a big slow query slowing down the side, but we have thousands of small ones? Is it possible to share your profiler findings? I think that they might be useful for the future.

Yes, that's right. Before the fix we had an sql query (returning 20,000 records for a course where there are 20,000 users enrolled) and an additional SQL query for every one of these records. After the fix, we still have the big SQL query (not problematic) but we are no longer doing the 20,000 calls.

Before:
![image](https://user-images.githubusercontent.com/799065/176669982-5b86d136-f499-4e92-bf33-27ebd4c29384.png)

After:
![image](https://user-images.githubusercontent.com/799065/176670070-8cf9adfc-e954-4cfb-886e-cac1bd7b5672.png)

In the above image you can see the amount of calls to the `get_term` method which is a good indicator of what was going on.

Some extra details:
- There were three reasons why all the terms were being requested: `lazy_load_term_meta`, `cache_results` and the call to `has_term(...)`.
- The first two basically assumed that for any WPQuery to a post it is a good idea to cache all the information regarding its taxonomies and terms. In our case, this is too much and never used.
- The last one (`has_term(...)`) has to do with the way it is implemented. Probably because it is assumed that terms would have been cached because of the above reason.


## Profiling statistics for posterity:

### Before the fix:
![image](https://user-images.githubusercontent.com/799065/176671661-efb4bba3-8b69-4c6d-8d79-060fd1cf6cd6.png)
[cachegrind.out.9938.trunk.zip](https://github.com/Automattic/sensei/files/9019561/cachegrind.out.9938.trunk.zip)



### After the fix:
![image](https://user-images.githubusercontent.com/799065/176671728-2ee1f920-0a05-4410-9f94-babab9c8a842.png)
[cachegrind.out.9939.fixed.zip](https://github.com/Automattic/sensei/files/9019562/cachegrind.out.9939.fixed.zip)
